### DISCUSSION TITLE
Does anyone know how to address the problem in using eeg_group_by()?

### DISCUSSION THREAD

**Discussion by ghzhlchen** [bnicenboim/eeguana#238] [Q&A] [answered]
I have been using eeguana to plot ERPs. I always use the same scripts but modify the attributes. Now I have a problem in using eeg_group_by(). 
![image](https://github.com/bnicenboim/eeguana/assets/131355522/b0a55d57-add5-49bf-85ba-30484b295565)
![image](https://github.com/bnicenboim/eeguana/assets/131355522/bc8f6ab0-c3ed-48a8-966b-3b3a02d11ca5)
The .Signal is problematic after grouping. 
When I chech the original .Signal, it looks normal. 
![image](https://github.com/bnicenboim/eeguana/assets/131355522/6243411a-de43-4a5b-880b-647a0f02f833)

Thanks for your help!

**Comment by bnicenboim** [comment]
Hi,
How do you know it's problematic? What were you expecting?
What was working before that now doesn't?

**Reply by ghzhlchen** [reply]
Hi!
I was trying to plot the ERPs in their layouts. I got the following information. 
![image](https://github.com/bnicenboim/eeguana/assets/131355522/a76c804d-eff7-414c-999a-a4f54239e0c7)
I guess it is because the grouping is problematic. 
A successful grouping should be like
![image](https://github.com/bnicenboim/eeguana/assets/131355522/22770f72-96fb-42c3-a826-e3887473d369)
Thank you

**Comment by bnicenboim** [comment]
Could you show me how `predicate` looks like? Or even better upload (maybe a subset of the data)?
Also please post a reproducible example? if you install,`reprex `(https://reprex.tidyverse.org/), you can use the Addins menu in Rstudio and do Reprex selection.

**Reply by ghzhlchen** [reply]
I have tried using the reprex package but it does not work because it cannot load/recognize the dataset (i.e., "ERROR: object 'predicate' not found"), I do not know if I have permission to share (a subset) of the dataset, but if that would be necessary I can ask my supervisors.

Meanwhile, I'll try to answer your questions as best I can: predicate is simply the name I've given my EEG data; it's an eeg_lst type object containing the .signal, .event and .segments lists.
I have done some further testing, and the issue seems to be with the eeg_summarize(across_ch(mean, na.rm = TRUE)) function. If I only apply the grouping and examine the resulting dataframe, I get a grouped dataframe that still has all the correct sampled data (integer ranging from -511 to 102). However, if I apply the eeg_summarize function, only the first sample point (-511) is kept for which the summaries are then made, which leads to the problem that every group only has one observation.
In a different script I use the exact same code, and there it does work as intended and all the unique sample points are kept, so I do not know why it doesn't work now.

**Reply by bnicenboim** [reply]
I know that predicate is the EEG data, but I don't know its structure. Can you reproduce your problem with another free dataset? Could you use `data_faces_ERPs` to reproduce the problem? or just create fake data?
Without a reproducible example (and code that I can copy paste and try in my computer) I cannot do much.

**Comment by ghzhlchen** [comment]
[Uploading Working.zip…]()
[Uploading Not_working.zip…]()

Thank you very much for your patience. I've asked the permission to share a subset of the data. I tried to attached the scripts and data for a situation where the code does work as expected ('Working' folder) and also the problematic one that I mentioned when I opened this question ('Not_working' folder). But I am not sure if the uploading works.

**Reply by bnicenboim** [reply]
it doesn't seem to be working. Maybe use a link to google drive or something that allows for large files (if they are large.) But I rather see the code in the post as code (not as an image) so that it can be useful to other people (after the links stop working).

**Reply by ghzhlchen** [reply]
Thank you very much. Here is the link
https://drive.google.com/file/d/1xwxwNAgyjF_RuOg88tOPYVnTOJNRA8At/view?usp=drive_link 
the codes as following

```{r} 
ERP_predicate <- predicate %>%
  eeg_group_by(.sample, predicate_congruency ) %>%
  eeg_summarize(across_ch(mean, na.rm = TRUE))

predicate_plot <- ERP_predicate %>%
  ggplot(aes(x = .time + 0.8, y = .value)) +  
  geom_line(aes(color = predicate_congruency)) +
  facet_wrap(~.key) +
  theme(legend.position = "bottom") +
  ggtitle("ERPs for congruent and incongruent predicates") +
  theme_eeguana()
predicate_plot %>% plot_in_layout()
```

**Reply by bnicenboim** [reply]
Hi, I'm quite busy at the moment. It'll take me at least couple of weeks to look at it. Maybe the eeguana expert, @stonekate, can take a look in the meanwhile?
Otherwise, you'll need a bit of patience....

**Reply by stonekate** [reply]
I took a quick look at the ReprEx (thanks) and it is indeed quite odd that one eeg_group_by groups by .sample and the other (identical) seems to be ignoring the .sample argument and grouping across it... I'll look more deeply this week and see if I can come up with anything!

**Reply by ghzhlchen** [reply]
I updated the package to the latest version as advised by Kate. However, the problem is still there. So, we can exclude the version factor now. 
![image](https://github.com/bnicenboim/eeguana/assets/131355522/90e2faa6-0645-4c29-b2ef-9d12269991de)

**Reply by stonekate** [reply]
Hi Zhenlin - so I still don't really know what the problem is, but I found a workaround!

I noticed in your "working" script that you did a filter() before you do the group_by(). If you add a filter before your group_by() in the "not working" script, then the grouping works. I just added: `filter(predicate_congruency %in% c("incongruent","congruent"))`. This returns all the rows so the data is the same. I don't know why this should make group_by() now recognise the .sample in the .signal table.... maybe @bnicenboim will have some idea.

Note that filter() works with just the .segments table variable name (i.e. without specifying predicate$.segments$predicate_congruency). In fact, this might be the safer way to do it - I had wondered whether specifying predicate$.segments$RC_type in the "working" script just returned the .segments table and maybe that was part of the difference between the two scripts. I don't think so, as the result of filter() was the same either way, but I think generally probably it's safer just to give the variable name in case there are cases where this does happen. 

Best,
Kate

**Reply by bnicenboim** [reply]
ohh, so there's a bug... :(
@stonekate, can you summarize the issue (and the workaround) in a github issue so that I can fix it?

**Reply by stonekate** [reply]
mmm I don't think it's a bug - I used some of your test code @bnicenboim but I can't reproduce the issue. I think I may have found the issue and it's specific to Zhenlin's code (I'm in the lab at the moment but will test something out and post it soon).

**Reply by stonekate** [reply, marked as answer]
Ok @ghzhlchen, the problem was actually in both scripts, it's just that the `filter()` in the "working" script somehow disguised it. I'll use the line numbers from the "working" script: The issue was the chunks beginning at lines 29 and 49. For reasons described [here](https://bruno.nicenboim.me/eeguana/articles/gotchas.html), it's better not to edit the .segments tables directly. A safer version of your script would be:

```
# Load multiple subjects using purrr::map, extracting subject IDs from file names.
COM_list <- purrr::map(list.files("./", "vhdr"), ~
read_vhdr(.x))
# Bind all the eeg_lsts into a large one:
COM <- eeg_bind(COM_list)

#add condition and trial to .segment
events_table <- COM$.events

# generate .trial and .condition columns
result <- events_table %>%
  group_by(.id) %>%
  mutate(
    .trial = .description[which(.type == "Stimulus")[1]],
    .condition = .description[which(.type == "Stimulus")[2]]
  ) %>%
  slice(1) %>%
  select(.id, .trial, .condition) %>%
  ungroup()

COM <- COM %>%
  left_join(., result %>% select(.id, .trial, .condition), by = ".id")

#left join RC_type
# read the coding table
RCtype_coding <- read.csv("RCtype_coding.csv")
RCtype_coding

# left join the coding table to the DEM list and rename type_exclude_coding
COM <- left_join(COM, select(RCtype_coding, .trial, type_exclude_coding), by = ".trial") %>%
  rename(RC_type = type_exclude_coding)

#grand-averaged erps
ERP_COM <- COM %>%
  filter(RC_type != "EX") %>%   # Exclude rows where type_exclude_coding is "EX"
  eeg_group_by(.sample, RC_type) %>%
  eeg_summarize(across_ch(mean, na.rm = TRUE))

COM_plot <- ERP_COM %>%
  ggplot(aes(x = .time, y = .value)) +  # i removed the addition of 1600 ms 
  geom_line(aes(color = RC_type)) +
  facet_wrap(~.key) +
  theme(legend.position = "bottom") +
  ggtitle("ERPs for SR and OR on COMs") +
   # Adjust the limits as needed
  # scale_x_continuous(limits = c(-0.2, 1.0)) + # this now doesn't suit the data
  theme_eeguana()
COM_plot %>% plot_in_layout()
```

This recovers the same plot you had before, although the time scale is now a little different - I'm not sure what it should be, I'll leave that one for you! You could consider segmenting before adding the .trial and .condition columns, something like `COM %>% eeg_segment(.type == "New Segment", .lim = c(0, 2))`. This might give you more control over where time zero is, but that really depends on how your data is. Anyway, the code above should also be applicable for your "not working" example.

**Reply by ghzhlchen** [reply]
@stonekate I realized I forgot to reply. Thank you very much for your help.
